shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java

```diff
@@ -80,25 +80,25 @@ public void onRemoval(RemovalNotification<String, Gauge> removal) {
                 public void onRemoval(RemovalNotification<String, Meter> removal) {
                     Meter meter = removal.getValue();
                     if (removal.getCause() != RemovalCause.REPLACED) {
-                        Exceptions.checkNotNullOrEmpty(meter.getName(), "meter");
-                        metrics.remove(meter.getName());
-                        log.debug("Removed Meter: {}.", meter.getName());
+                        metrics.remove(meter.getId());
+                        log.debug("Removed Meter: {}.", meter.getId());
                     }
                 }
             }).
             build();
     }
 
     @Override
-    public void incCounterValue(String name, long delta) {
+    public void incCounterValue(String name, long delta, String... tags) {
         Exceptions.checkNotNullOrEmpty(name, "name");
         Preconditions.checkNotNull(delta);
-        String counterName = name + ".Counter";
+        String counterName = nameFromTags(name, tags) + ".Counter";
```

Comment:
This is yet another example of inefficient code on the hot path. You are concatenating two strings, which requires a new allocation and memory copy. Consider updating `nameFromTags` to accept additional args to format the string from the first pass.

Reply:
Also you only need this variable in one place, and subject to a condition. Please inline this in the `createCounter` call below (line 101).

Reply:
On third thought, doesn't `nameFromTags(name, tags)` return `name` if `tags` is null or empty? So your conditional at line 101 will always return `name`.

Reply:
Please have a look at the MetricKey implementation.